In [7]:
import pandas as pd
import numpy as np
import re

In [3]:
df = pd.read_csv('GSAF5.csv', encoding = 'latin1')

In [5]:
df.sample(frac=1).head()

Case Number                       Date  Year        Type  \
3744  1960.08.22.R.       Reported 22-Aug-1960  1960  Unprovoked   
5989        ND.0003                  1900-1905     0  Unprovoked   
4717     1934.02.24                  24-Feb-34  1934  Unprovoked   
2531   1988.01.05.R  Reported      05-Jan-1988  1988    Provoked   
828      2010.01.12                  12-Jan-10  2010  Unprovoked   

               Country                   Area          Location  \
3744  PAPUA NEW GUINEA     Milne Bay Province               NaN   
5989               USA         North Carolina    Ocracoke Inlet   
4717         AUSTRALIA          Torres Strait  Adolphus Channel   
2531      SOUTH AFRICA  Western Cape Province  Great Brak River   
828       SOUTH AFRICA       Western Province         Fish Hoek   

                                     Activity  \
3744                              Free-diving   
5989                                 Swimming   
4717                      Splashing in water    
2531  Returning to shore, collided with shark   
828                                  Standing   

                                                   Name Sex   ...  \
3744                                         "a native"    M  ...   
5989                              Coast Guard personnel    M  ...   
4717                               Rixon, an aboriginal    M  ...   
2531  5 m inflatable boat, occupants: Kobus Potgiete...  NaN  ...   
828                                       Lloyd Skinner    M  ...   

                                   Species   \
3744                                    NaN   
5989                                    NaN   
4717  3.4 m [11']  tiger shark & a 5' shark   
2531                              6 m shark   
828                             White shark   

                    Investigator or Source                             pdf  \
3744                    The Age, 8/22/1960      1960.08.22.R-Milne-PNG.pdf   
5989  F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
4717          V.M. Coppleson (1958), p.243            1934.02.24-Rixon.pdf   
2531               Natal Mercury, 1/5/1988   1988.01.05-Potgeiter-boat.pdf   
828        L. Cohen, Times Live, 1/12/2010          2010.01.12-Skinner.pdf   

                                           href formula  \
3744  http://sharkattackfile.net/spreadsheets/pdf_di...   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...   
4717  http://sharkattackfile.net/spreadsheets/pdf_di...   
2531  http://sharkattackfile.net/spreadsheets/pdf_di...   
828   http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  Case Number.1  \
3744  http://sharkattackfile.net/spreadsheets/pdf_di...  1960.08.22.R.   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...        ND.0003   
4717  http://sharkattackfile.net/spreadsheets/pdf_di...     1934.02.24   
2531  http://sharkattackfile.net/spreadsheets/pdf_di...   1988.01.05.R   
828   http://sharkattackfile.net/spreadsheets/pdf_di...     2010.01.12   

      Case Number.2 original order Unnamed: 22 Unnamed: 23  
3744  1960.08.22.R.           2249         NaN         NaN  
5989        ND.0003              4         NaN         NaN  
4717     1934.02.24           1276         NaN         NaN  
2531   1988.01.05.R           3463         NaN         NaN  
828      2010.01.12           5165         NaN         NaN  

[5 rows x 24 columns]

In [6]:
df.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [8]:
df['Date'].value_counts().sort_index()

    10-Jan-2009    1
    15-Jun-1937    1
    16-Jan-1970    1
    22-Jul-2013    1
   21-Sep-1908     1
                  ..
Summer-2008        1
Winter 1942        1
Winter 1969        1
Woirld War II      1
World War II       2
Name: Date, Length: 5128, dtype: int64

In [9]:
df.loc[0,'Date']

'18-Sep-16'

In [10]:
pd.to_datetime(df.loc[0,'Date'],infer_datetime_format = True)

Timestamp('2016-09-18 00:00:00')

In [11]:
def are_you_inferable(date):
    if pd.to_datetime(date, infer_datetime_format = True):
        return True
    else:
        False

In [12]:
are_you_inferable('18-Sep-16')

True

In [15]:
df['Date'].apply(are_you_inferable)

ValueError: ('Unknown string format:', 'Reported  14-Jul-2016')

This is not working

In [16]:
def are_you_inferable2(date):
    try:
        pd.to_datetime(date, infer_datetime_format = True)
        return True
    except:
        return False

In [17]:
are_you_inferable2('Reported  14-Jul-2016')

False

In [19]:
df['Date'].apply(are_you_inferable2)

0        True
1        True
2        True
3        True
4        True
        ...  
5987    False
5988    False
5989    False
5990    False
5991    False
Name: Date, Length: 5992, dtype: bool

In [20]:
df['Date'].apply(are_you_inferable2).value_counts()

True     5135
False     857
Name: Date, dtype: int64

Move the 857 rows that could not be infered into a separated df for further inspection

In [23]:
df_ninfer = df[df['Date'].apply(are_you_inferable2) == False]
df_ninfer.head()

Case Number                   Date  Year        Type         Country  \
33   2016.07.14.4  Reported  14-Jul-2016  2016  Unprovoked         BAHAMAS   
34   2016.07.08.R   Reported 08-Jul-2016  2016  Unprovoked           SPAIN   
85   2016.03.03.R   Reported 03-Mar-2016  2016  Unprovoked       AUSTRALIA   
90   2016.02.10.R   Reported 10-Feb-2016  2016     Invalid  CAYMAN ISLANDS   
100  2016.01.11.R   Reported 11-Jan-2016  2016  Unprovoked       AUSTRALIA   

                Area                       Location            Activity  \
33               NaN                    Tiger Beach        Scuba Diving   
34    Canary Islands        Las Teresitas, Tenerife              Wading   
85   South Australia                    Wrights Bay             Fishing   
90      Grand Cayman              Stingray City Bar  Feeding stingrays?   
100       Queensland  Happy Valley Beach, Caloundra             Surfing   

                  Name Sex   ...              Species   \
33   Michael Dornellas    M  ...       Lemon shark, 9'   
34              female    F  ...           Angel shark   
85          Lee Taplin    M  ...         Bronze whaler   
90     Richard Branson    M  ...  No shark involvement   
100       Shane Hilder    M  ...       Wobbegong shark   

            Investigator or Source                                pdf  \
33              GrindTV, 7/14/2016        2016.07.14.R-TigerBeach.pdf   
34          La Opinion de Tenerife             2016.07.08.R-Spain.pdf   
85                9 News, 3/1/2016            2016.03.03.R-Taplin.pdf   
90                      R. Branson  2016.02.10.R-Branson-stingray.pdf   
100  ABC Sunshine Coast, 1/11/2016            2016.01.11.R-Hilder.pdf   

                                          href formula  \
33   http://sharkattackfile.net/spreadsheets/pdf_di...   
34   http://sharkattackfile.net/spreadsheets/pdf_di...   
85   http://sharkattackfile.net/spreadsheets/pdf_di...   
90   http://sharkattackfile.net/spreadsheets/pdf_di...   
100  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                  href Case Number.1  \
33   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.07.14.R   
34   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.07.08.R   
85   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.03.03.R   
90   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.02.10.R   
100  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.01.11.R   

    Case Number.2 original order Unnamed: 22 Unnamed: 23  
33   2016.07.14.4           5960         NaN         NaN  
34   2016.07.08.R           5959         NaN         NaN  
85   2016.03.03.R           5908         NaN         NaN  
90   2016.02.10.R           5903         NaN         NaN  
100  2016.01.11.R           5893         NaN         NaN  

[5 rows x 24 columns]

In [24]:
df_ninfer.shape

(857, 24)

In [25]:
df_ninfer['Date'].value_counts()

No date                 6
No date, Before 1963    5
1970s                   5
Before 1958             4
Reported 10-Oct-1906    4
                       ..
Reportd 15-Jul-1894     1
Reported 07-Mar-1930    1
Reported 06-Mar-1886    1
Reported 15-Dec-1909    1
1844.07.16.R            1
Name: Date, Length: 771, dtype: int64

In [30]:
date_pattern = r'(\d{4}\.\d{2}\.\d{2})'
df_ninfer['Case Number'].str.extract(date_pattern, expand = False)

33      2016.07.14
34      2016.07.08
85      2016.03.03
90      2016.02.10
100     2016.01.11
           ...    
5987           NaN
5988           NaN
5989           NaN
5990           NaN
5991           NaN
Name: Case Number, Length: 857, dtype: object

In [31]:
df_ninfer['Case Number'].str.extract(date_pattern, expand = False).apply(are_you_inferable2) == False

33      False
34      False
85      False
90      False
100     False
        ...  
5987    False
5988    False
5989    False
5990    False
5991    False
Name: Case Number, Length: 857, dtype: bool